### ISSUE TITLE
MA0004 should not warn, when a configured (Value)Task is awaited

### ISSUE THREAD

**Issue by Pretasoc** [meziantou/Meziantou.Analyzer#232]
**Context**
Rule identifier: _MA0004_
Target Framework: _.NET 5.0_
Language version: _9.0_

**Describe the bug**
When using `IAsyncEnumerable<T>` in a `foreach` loop, i prefer ist so set the cancellation token and configure the await before the loop, because i find it easier to read. Now a MA0004 is triggered, although the await _is_ configured.

**To Reproduce**
Code or steps to reproduce the behavior:

````c#
ConfiguredCancelableAsyncEnumerable<Foo> foos= GetFooAsync()
    .WithCancellation(cancellation)
    .ConfigureAwait(false);

await foreach (var foo in foos)
{
    // do stuff
}
````

**Comment by meziantou** [comment]
To correctly handle this case, the rule should do a dataflow analysis to understand how the variable is defined. This is not trivial and adds complexity. Instead, I think it would be ok to skip the case where the collection is a variable reference and the variable is a `ConfiguredCancelableAsyncEnumerable`. However, this won't validate that you actually call `ConfigureAwait`.

````c#
ConfiguredCancelableAsyncEnumerable<Foo> foos = GetFooAsync()
    .WithCancellation(cancellation)
    .ConfigureAwait(false);

await foreach (var foo in foos) { } // ok
````

````c#
ConfiguredCancelableAsyncEnumerable<Foo> foos = GetFooAsync()
    .WithCancellation(cancellation);

await foreach (var foo in foos) { } // ok, even if there is no ConfigureAwait
````

````c#
IAsyncEnumerable<Foo> foos= GetFooAsync();

await foreach (var foo in foos) { } // report diagnostic
````

Would it be ok for you?

**Comment by Pretasoc** [comment]
For me no flow analysis would be neccessary. Just allowing `ConfiguredCancelableAsyncEnumerable` would work fine for me.

**Comment by meziantou** [comment]
The next version (>=1.0.653) should fix this issue
